mcs/tools/wasm-tuner/tuner.cs

```diff
@@ -76,12 +89,94 @@ public class WasmTuner
 		}
 	}
 
+	void Usage () {
+		Console.WriteLine ("Usage: tuner.exe <arguments>");
+		Console.WriteLine ("Arguments:");
+		Console.WriteLine ("--gen-icall-table icall-table.json <assemblies>.");
+		Console.WriteLine ("--gen-pinvoke-table <list of native library names separated by commas> <assemblies>.");
+	}
+
 	int Run (String[] args) {
-		if (args.Length < 3 || args [0] != "--gen-icall-table") {
-			Console.WriteLine ("Usage: tuner.exe --gen-icall-table icall-table.json <assemblies>.");
+		if (args.Length < 1) {
+			Usage ();
 			return 1;
 		}
-		return GenIcallTable (args);
+		string cmd = args [0];
+		if (cmd == "--gen-icall-table") {
+			if (args.Length < 3) {
+				Usage ();
+				return 1;
+			}
+			return GenIcallTable (args);
+		} else if (cmd == "--gen-pinvoke-table") {
+			return GenPinvokeTable (args);
+		} else {
+			Usage ();
+			return 1;
+		}
+	}
+
+	int GenPinvokeTable (String[] args) {
+		var modules = new Dictionary<string, string> ();
+		foreach (var module in args [1].Split (','))
+			modules [module] = module;
+
+		args = args.Skip (2).ToArray ();
+		pinvokes = new List<Pinvoke> ();
+		foreach (var fname in args) {
+			var a = AssemblyDefinition.ReadAssembly (fname);
+
+			foreach (var type in a.MainModule.Types) {
+				ProcessTypeForPinvoke (type);
+				foreach (var nested in type.NestedTypes)
+					ProcessTypeForPinvoke (nested);
+			}
+		}
+
+		Console.WriteLine ("// GENERATED FILE, DO NOT MODIFY");
+		Console.WriteLine ("typedef struct {");
+		Console.WriteLine ("const char *name;");
+		Console.WriteLine ("void *func;");
+		Console.WriteLine ("} PinvokeImport;");
+		Console.WriteLine ();
+
+		foreach (var pinvoke in pinvokes) {
+			if (modules.ContainsKey (pinvoke.Module))
+				Console.WriteLine ($"void {pinvoke.EntryPoint} ();");
+		}
+
+		foreach (var module in modules.Keys) {
+			string symbol = module.Replace (".", "_") + "_imports";
+			Console.WriteLine ("static PinvokeImport " + symbol + " [] = {");
+			foreach (var pinvoke in pinvokes) {
+				if (pinvoke.Module == module)
+					Console.WriteLine ("{\"" + pinvoke.EntryPoint + "\", " + pinvoke.EntryPoint + "},");
+			}
+			Console.WriteLine ("{NULL, NULL}");
+			Console.WriteLine ("};");
+		}
+		Console.Write ("static void *pinvoke_tables[] = { ");
```

Comment:
Nit, ignorable, you know, non-`void*` preferable, and `const`. `const` provides efficiency in most C/C++ but maybe not in wasm.